libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc

```diff
@@ -0,0 +1,528 @@
+    // which can be written as
+    //   (beta * input) * input = output - gamma
+    this->pb.add_r1cs_constraint(
+        {input * beta, input, output - gamma},
+        FMT(this->annotation_prefix, " beta * x = y - gamma"));
+}
+
+// compute a witness y for a given input x for the computation y =
+// beta x^2 + gamma, where x=input, y=output
+template<typename FieldT, size_t generator>
+void flystel_Q_gamma_prime_field_gadget<FieldT, generator>::
+    generate_r1cs_witness()
+{
+    const FieldT input_value =
+        input.evaluate(this->pb.full_variable_assignment());
+    // y = beta x^2 + gamma
+    this->pb.val(output) = this->beta * input_value * input_value + this->gamma;
+
+    printf("[%s:%d] flystel_Q_gamma_prime_field_gadget\n", __FILE__, __LINE__);
+    printf("[%s:%d] %s() input ", __FILE__, __LINE__, __FUNCTION__);
+    input_value.print();
+    printf("[%s:%d] %s() output ", __FILE__, __LINE__, __FUNCTION__);
+    this->pb.val(output).print();
+}
+
+template<typename FieldT, size_t generator>
+flystel_Q_delta_prime_field_gadget<FieldT, generator>::
+    flystel_Q_delta_prime_field_gadget(
+        protoboard<FieldT> &pb,
+        const linear_combination<FieldT> &input,
+        const pb_variable<FieldT> &output,
+        const std::string &annotation_prefix)
+    : gadget<FieldT>(pb, annotation_prefix)
+#ifdef FLYSTEL_DEBUG
+    , beta(DEBUG_FLYSTEL_BETA)
+    , delta(DEBUG_FLYSTEL_DELTA)
+#elif
+    , beta(FieldT(generator))
+    , delta(FieldT(generator).inverse())
+#endif // #ifdef FLYSTEL_DEBUG
+    , input(input)
+    , output(output)
+{
+}
+
+template<typename FieldT, size_t generator>
+void flystel_Q_delta_prime_field_gadget<FieldT, generator>::
+    generate_r1cs_constraints()
+{
+    // Constraint has the form:
+    //   beta * input^2 + delta = output
+    // which can be written as
+    //   (beta * input) * input = output - delta
+    this->pb.add_r1cs_constraint(
+        {input * beta, input, output - delta},
+        FMT(this->annotation_prefix, " beta * x = y - delta"));
+}
+
+// compute a witness y for a given input x for the computation y =
+// beta x^2 + delta, where x=input, y=output
+template<typename FieldT, size_t generator>
+void flystel_Q_delta_prime_field_gadget<FieldT, generator>::
+    generate_r1cs_witness()
+{
+    const FieldT input_value =
+        input.evaluate(this->pb.full_variable_assignment());
+    // y = beta x^2 + delta
+    this->pb.val(output) = this->beta * input_value * input_value + this->delta;
+
+    printf("[%s:%d] flystel_Q_delta_prime_field_gadget\n", __FILE__, __LINE__);
+    printf("[%s:%d] %s() input ", __FILE__, __LINE__, __FUNCTION__);
+    input_value.print();
+    printf("[%s:%d] %s() output ", __FILE__, __LINE__, __FUNCTION__);
+    this->pb.val(output).print();
+}
+
+// R1CS constraints for the operation y = beta x^3 + gamma with
+// x=input, y=output. This operation is represented with two
+// multiplications as y-gamma = ((beta x * x) * x). Equivalently:
```

Comment:
Let's tighten up the language here.  There are 3 multiplications in the expression (multiplication by gamma), but we rearrange the expression to write this in terms of two constraints.

Reply:
Is the only correction here that there are 3 and not 2 multiplications, or should the comment also be revised to be more concise?

Reply:
Technically, but I would not mention the "3 multiplications", since they aren't really relevant (we could have up to `3*n` "multiplications" for every constraint, for a witness of size `n`, since each leg of a constraint is a linear combination of witness members).

We have an expression `y = beta x^3 + gamma` and this gadget uses two constraints to implement this.

Reply:
Addressed.